Discovery/Discovery/Presentation/WebPrograms/ProgramWebviewViewModel.swift

```diff
@@ -0,0 +1,239 @@
+                errorMessage = CoreLocalization.Error.slowOrNoInternetConnection
+            } else {
+                errorMessage = CoreLocalization.Error.unknownError
+            }
+        }
+    }
+    
+    private var courseState: CourseState? {
+        guard courseDetails?.isEnrolled == false else { return nil }
+        
+        if let enrollmentStart = courseDetails?.enrollmentStart, let enrollmentEnd = courseDetails?.enrollmentEnd {
+            let enrollmentsRange = DateInterval(start: enrollmentStart, end: enrollmentEnd)
+            if enrollmentsRange.contains(Date()) {
+                return .enrollOpen
+            } else {
+                return .enrollClose
+            }
+        } else {
+            return .enrollOpen
+        }
+    }
+}
+
+extension ProgramWebviewViewModel: WebViewNavigationDelegate {
+    @MainActor
+    public func webView(
+        _ webView: WKWebView,
+        shouldLoad request: URLRequest,
+        navigationAction: WKNavigationAction
+    ) -> Bool {
+        guard let URL = request.url else { return false }
+        
+        if let urlAction = urlAction(from: URL),
+           handleNavigation(url: URL, urlAction: urlAction) {
+            return true
+        }
+        
+        let capturedLink = navigationAction.navigationType == .linkActivated
+        let outsideLink = (request.mainDocumentURL?.host != self.request?.url?.host)
+        var externalLink = false
+        
+        if let queryParameters = request.url?.queryParameters,
+            let externalLinkValue = queryParameters["external_link"] as? String,
+           externalLinkValue.caseInsensitiveCompare("true") == .orderedSame {
+            externalLink = true
+        }
+        
+        if let url = request.url, outsideLink || capturedLink || externalLink, UIApplication.shared.canOpenURL(url) {
+            DispatchQueue.main.async { [weak self] in
+                self?.router.presentAlert(
+                    alertTitle: DiscoveryLocalization.Alert.leavingAppTitle,
+                    alertMessage: DiscoveryLocalization.Alert.leavingAppMessage,
+                    positiveAction: CoreLocalization.Webview.Alert.continue,
+                    onCloseTapped: {
+                        self?.router.dismiss(animated: true)
+                    }, okTapped: {
+                        UIApplication.shared.open(url, options: [:])
+                    }, type: .default(positiveAction: CoreLocalization.Webview.Alert.continue)
+                )
+            }
+            return true
+        }
+        
+        return false
+    }
+    
+    private func urlAction(from url: URL) -> WebviewActions? {
+        guard isValidAppURLScheme(url),
+                let url = WebviewActions(rawValue: url.appURLHost) else { return nil }
+        return url
+    }
+    
+    @MainActor
+    private  func handleNavigation(url: URL, urlAction: WebviewActions) -> Bool {
```

Comment:
Extra space
`private  func` -> `private func`

Reply:
Thank you!
I found another one in DiscoveryWebviewViewViewModel #161
`private func handleNavigation` -> `private func handleNavigation`.
Could you please fix this one as well?

Reply:
Done

Reply:
Thank you!